lib/engine/game/g_18_esp/step/tracker.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+require_relative '../../../step/tracker'
+
+module Engine
+  module Game
+    module G18ESP
+      module Tracker
+        include Engine::Step::Tracker
+
+        def setup
+          super
+          @round.extra_mine_lay = false
+          @round.mine_tile_laid = false
+        end
+
+        def lay_tile_action(action)
+          old_tile = action.tile
+          super
+
+          if @game.mine_hex?(action.hex) && old_tile.color == :white
+            @round.num_laid_track -= 1
+            @round.mine_tile_laid = true
+          end
+
+          action.entity.goal_reached!(:destination) if @game.check_for_destination_connection(action.entity)
+        end
+
+        def extra_cost(tile, tile_lay, hex)
+          @game.mine_hex?(hex) ? 0 : super
+        end
+
+        def round_state
+          super.merge(
+            {
+              extra_mine_lay: false,
+              mine_tile_laid: false,
+            }
+          )
+        end
+
+        def tracker_available_hex(entity, hex)
+          get_tile_lay(entity)
+          return super && @game.mine_hex?(hex) if @round.extra_mine_lay
+
+          @round.mine_tile_laid ? super && !@game.mine_hex?(hex) : super
+        end
+
+        def get_tile_lay(entity)
+          action = super
+          # if action is nil, and mine wasn't laid, grant a lay action buy only for mine
+          if action.nil? && !@round.mine_tile_laid
+            @round.extra_mine_lay = true
+            return { lay: true, upgrade: false, cost: 0 }
+          end
+          action
+        end
+
+        def potential_tiles(entity, hex)
+          tiles = super
+
+          if @game.north_hex?(hex)
+            tiles.reject! { |tile| tile.paths.any? { |path| path.track == :broad } }
+          else
+            tiles.reject! { |tile| tile.paths.any? { |path| path.track == :narrow } }
+          end
+          tiles
+        end
+
+        def legal_tile_rotation?(entity, hex, tile)
+          hex.tile.towns.none?(&:halt?) ? super : halt_upgrade_legal_rotation?(entity, hex, tile)
+        end
+
+        def halt_upgrade_legal_rotation?(entity_or_entities, hex, tile)
+          # entity_or_entities is an array when combining private company abilities
+          entities = Array(entity_or_entities)
```

Comment:
a corp can destory a halt (aka mines), to a regular town in green for a fee.
This method handles the upgrade logic needed to do that